Models branch

IvanAndrau opened the pull request:
Created all classes except for User;

We need to decide what type of Id we want to use [int] or [Guid]
`        [Key]
        public int Id { get; set; }     //could be [Guid]`
In my opinion we should use [Guid] because it's more secure and it's more difficult to guess it, also it cannot be duplicated. 
I did not create User table because I think we should use Identity framework to create them. User model from identity framework has all the properties as our class except for 'CreatedAt' prop, but we can add it by overriding the provided user class.

IvanAndrau committed:
Created a branch for models

IvanAndrau committed:
Feat: Add JobApplication model & Remove User model

IvanAndrau committed:
Fix: Move back Models folder to <ItemGroup>

IvanAndrau committed:
Feat: Add Interview table

IvanAndrau committed:
Feat: Add EmploymentType model

IvanAndrau committed:
Feat: Add Status table

IvanAndrau committed:
Feat: Add JobApplicationHistory table model
Don't get confused by model and table words in each Commit, I mixed them up accidentally. All model and table words mean class

LarryAndrau commented on jobApplicationTracker/Models/EmploymentType.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public class EmploymentType
+    {
+        [Key]
```
I'd prefer floating API instead of attributes, since it's more flexible. But it's up to you

eric-muganga replied:
Thanks for the suggestion, Larry! A floating API is new to me. Could you elaborate more on how to effectively implement it and integrate it into the project? I’d love to understand how it works in practice and how to best use it in this context. Your guidance would be greatly appreciated!

LarryAndrau commented on jobApplicationTracker/Models/Interview.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace jobApplicationTrackerApi.Models
+{
+    public class Interview
```
I do not see Navigation Properties. It's related to all models

eric-muganga replied:
Thanks for pointing that out, Larry! You're absolutely right we need navigation properties to simplify loading related data and allow querying of relationships. I realize I forgot to inform Ivan about this earlier. Appreciate the feedback! 👍

LarryAndrau commented on jobApplicationTracker/Models/Interview.cs:
```diff
@@ -0,0 +1,16 @@
+        [MaxLength(50, ErrorMessage = "Type should be less than 50 characters")]
+        public string Type { get; set; }
+        public DateTime? Date { get; set; }      // I made Date nullable in case you don't know the date of interview
+        public string Notes { get; set; } = string.Empty;   // should it be string.Empty or string? 
```
Are you going to limit length of string fields in DB? like nvarchar(400)

eric-muganga replied:
Thanks for bringing this up, Larry! Yes, we should limit the length of string fields in the database to ensure consistency and prevent potential storage issues.I'll review the models and update the annotations accordingly

LarryAndrau commented on jobApplicationTracker/Models/JobApplication.cs:
```diff
@@ -0,0 +1,27 @@
+        public DateTime? ApplicationDate { get; set; }   //class diagram want it to be not null, but if we are planning to create a canban board if it will be in a so called TODO table it should be null in my opinion. 
+        public DateTime? InterviewDate { get; set; }
+        public string? Notes { get; set; }
+        public decimal? Salary { get; set; }
```
Do you need Currency? SalaryType  (hourly/monthly) etc?

eric-muganga replied:
Great point, Larry! Adding **Currency** and **SalaryType** (e.g., _hourly_, _monthly_) could make the Salary field more descriptive and flexible. I'll update the model to include these fields so that salary data can be more precise. Thanks for the suggestion!

IvanAndrau committed:
Refactor: Replace all int IDs with Guid, replaced UserId from int to string
Reason for changing UserId: Identity framework provides User table with id of type string

LarryAndrau commented on jobApplicationTracker/Models/Status.cs:
```diff
@@ -0,0 +1,15 @@
+        [MaxLength(50, ErrorMessage = "Status name should not exceed 50 characters")]
+        public string Name { get; set; }
+        [Required]
+        public int SortOrder { get; set; }
```
What this field for?

eric-muganga replied:
The SortOrder field is used to define the sequence in which statuses appear in the workflow. For example, statuses like "To Apply," "Applied," "Interview Scheduled," etc., can be displayed in the correct order in a UI

IvanAndrau commented:
Just to be clear Identity framework provides user with Id of type string, but it is a Guid in string format

LarryAndrau commented on jobApplicationTracker/Models/JobApplication.cs:
```diff
@@ -0,0 +1,27 @@
+        public DateTime? InterviewDate { get; set; }
+        public string? Notes { get; set; }
+        public decimal? Salary { get; set; }
+        public int EmploymentTypeId { get; set; }   //could be [Guid]
```
For EmploymentTypeId I do not see EmploymentType table. If it supposed to be enum in code - it should be int

eric-muganga replied:
We actually have the EmploymentTypes table defined at the top of the schema. It’s used to store predefined types like Umowa zlecenie and B2B, and EmploymentTypeId is a foreign key referencing that table. Let me know if you think using an enum instead of a table would be more appropriate for our use case!

LarryAndrau reviewed (commented):
You are creating data model.
In ideal case it should be Data model, Entity model and View model + mapping. It's better to start from Entity Model.

eric-muganga commented:
Thank you @LarryAndrau for highlighting the different types of models. While researching this topic, I’ve noticed a great deal of conflicting information online:

- Some sources describe **data models** as representations of the **business domain and associated logic or rules**.
- Others assert that **data models** simply mirror the **database structure**.

This discrepancy leads to confusion, which often extends to **entity models** as well.

However, the purpose of the **view model (DTO)** is generally clearer:

- It **controls what data is sent to the client**—be it a UI or an API consumer.
- It helps **prevent overexposure** of sensitive or unnecessary details.

**@LarryAndrau**, could you please share your opinion on how to best differentiate and use these model types?

IvanAndrau committed:
Feat: Rename folder Models to DataModels, Add FinancialInformation class, Add mapping properties, Change relationship of JobApplicationHistory class with Status table

IvanAndrau committed:
Feat: Add ViewModels folder with view models for each data model
